modelgym/gp_trainer.py

```diff
@@ -0,0 +1,122 @@
+import time
+
+import numpy as np
+from bson.son import SON
+from sklearn.model_selection import cross_val_score
+from skopt import gp_minimize
+
+import modelgym
+from modelgym.model import TASK_CLASSIFICATION
+
+
+class GPTrainer(object):
+    def __init__(self, n_estimators=5000, gp_evals=50, state=None, load_previous=False):
+        self.n_estimators, self.best_loss = n_estimators, np.inf
+        self.gp_evals, self.gp_eval_num = gp_evals, 0
+        self.default_params, self.best_params = None, None
+        self.best_n_estimators = None
+
+    def print_result(self, result, name='', extra_keys=None):
+        # TODO test
+        print('%s:\n' % name)
+        print('loss = %s' % (result['loss']))
+        if 'best_n_estimators' in result.keys():
+            print('best_n_estimators = %s' % result['best_n_estimators'])
+        elif 'n_estimators' in result.keys():
+            print('n_estimators = %s' % result['n_estimators'])
+        print('params = %s' % result['params'])
+        if extra_keys is not None:
+            for k in extra_keys:
+                if k in result:
+                    print("%s = %f" % (k, result[k]))
+
+    def crossval_optimize_params(self, model, cv_pairs, max_evals=None, verbose=True, batch_size=10,
+                                 tracker=None):
+        random_state = np.random.RandomState(1)
+        max_evals = max_evals or self.gp_evals
+        self.gp_eval_num, self.best_loss = 0, np.inf
+
+        prgp = modelgym.util.process_params_gp(model)
+        print("PRGP", prgp)
+
+        _ = gp_minimize(
+            func=lambda params: self.crossval_fit_eval(model=model, cv_pairs=cv_pairs, params=params, verbose=verbose),
+            dimensions=(prgp.values()), random_state=random_state, n_calls=max_evals,
+            n_jobs=max_evals - 1)
+
+        best_hyper_params = {k: v for k, v in zip(prgp.keys(), _.x)}
+        print(best_hyper_params)
+        bst = 1 - _.fun
+        print("Best accuracy score =", bst)
+        ans = best_hyper_params.copy()
+        ans['loss'] = bst
+        return ans if not isinstance(ans, SON) else ans.to_dict()
+
+    def crossval_fit_eval(self, model, cv_pairs, params=None, n_estimators=None, verbose=False):
+        params = params or model.default_params
+        if (isinstance(params, list)):
+            eta, max_depth, subsample, colsample_bytree, colsample_bylevel, min_child_weight, gamma, alpha, lambdax = \
+                params[0], params[1], params[2], params[3], params[4], params[5], params[6], params[7], params[8]
```

Comment:
hmm.. not sure this conversion should happen on trainer side. even worse - trainer should be parameter-agnostic. ask model to do the conversion.